README.md
# CBE Thermal Comfort Tool

A web interface for comfort model calculations and visualizations according to ASHRAE Standard-55, EN Standard 16798 and ISO Standard 7730. 

[Live deployment of the tool](http://comfort.cbe.berkeley.edu/).

## Getting Started

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes.

### Prerequisites

Python 3 and Node.js installed on your machine.

If you do not have Python installed on your machine you can follow [this guide](https://wiki.python.org/moin/BeginnersGuide/Download)

If you do not have Node.js installed on your machine you can follow [this guide](https://nodejs.org/en/download/)

### Installation

This guide is for Mac OSX, Linux or Windows.

1. **Get the source code from the GitHub repository**
```
$ git clone https://github.com/CenterForTheBuiltEnvironment/comfort-tool.git
$ cd comfort_tool
```
2. **Create a virtual environment using the following command:**

On Linux and MAC ` $ python3 -m venv venv `

On Windows ` py -3 -m venv venv `

3. **Activate the virtualenv:**

On Linux and MAC ` $ . venv/bin/activate `

On Windows ` venv\Scripts\activate `

Your shell prompt will change to show the name of the activated environment.

4. **Install Python dependencies**

The dependencies of the comfort tool are all contained in *requirements.txt*. 
Install them all using:
`$ pip install -r requirements.txt`

5. **Install Node.js dependencies**

`npm install`

6. **Run CBE Thermal Comfort Tool locally**

Now you should be ready to run the tool locally.
`python3 comfort.py`

Visit http://localhost:5000 in your browser to check it out. 
Note that whenever you want to run the tool, you have to activate the virtualenv first.

### Testing

We are using [Jest](https://jestjs.io/docs/en/getting-started.html) to test the JavaScript functions.

If you want to find out more please read their official documentation or look at how we are testing the [ERF functions (file name erf.js)](https://github.com/CenterForTheBuiltEnvironment/comfort_tool/blob/master/static/js/erf.js) using the test file `erf.test.js`.

Finally run `npm run test`. You should write tests for all the new functions you add and ensure that you get positive results from the tests. Also run tests before deploying a new version of the CBE Thermal Comfort Tool.

### Versioning
When you release a new version of the tool you should first use `bumpversion` to update the version of the tool. You can use the following command:
```cmd
bumpversion patch  # alternatively you can use minor or major instead of patch
```

Secondly you should describe the changes in `docs/changelog.md`

### Features

1. Models.
    * Adaptive (inputs: air temperature, MRT, mean outdoor temperature, air speed)
    * PMV with and without elevated air speed
2. Clothing ensemble creator.
3. Clothing and metabolic activity tables.
4. Dual units (SI, IP).
5. Local air speed control.
6. Operative temperature specification or Air temperature and Mean Radiant Temperature specification.
7. 1-to-1 humidity specification conversion (Relative humidity, vapor pressure, dew point, wet bulb, humidity ratio).
8. Globe temperature to MRT converter – input: air temperature, air speed, globe temperature, globe diameter, globe emissivity. Output: Mean Radiant Temperature.
9. Local thermal discomfort: short questionnaire that allows the user to assess risks of local thermal discomfort as defined by Standard 55.
10. Interactive charts: psychrometric chart, temperature–relative humidity chart, adaptive chart.
11. Mouse values: psychrometric variables displayed for each position of the cursor when the mouse moves on the chart.
12. Dynamic predictive clothing calculator.
13. Comfort Ranges: the ability to plot how the PMV and PPD ranges change in the psychrometric chart when clothing, metabolic activity, air speed, and mean radiant temperature variables are changed within a given range.
14. Comparison of comfort conditions: possibility of visually comparing two or more thermal comfort simulations. Instead of manually transcribing the output of thermal comfort simulations with the tool, it would be practical to enable a simple visualization to compare several conditions.
15. EN 15251 compliance: this section of the tool can be used to assess the compliance of indoor conditions to the European standard.


## Documentation
To edit or generate new documentation please refer to the [README.md](https://github.com/CenterForTheBuiltEnvironment/comfort_tool/blob/master/website/README.md) file which is located in the `./website` folder.

docs/changelog.md
---
id: changelog
title: CHANGELOG
sidebar_label: CHANGELOG
---

## Version 2.3.1 (2020-07-30)

Fix:

- Fix equation to calculate radiative heat transfer coefficient in SET equation

Features:

- Assuming that a person is sitting when calculating SET temperature

## Version 2.3.0 (2020-07-29)

Features:

- Added PHS model
- Added fan heatwaves model

## Version 2.2.0 (2020-07-14)

Fix:

- Error calculation of PMV in EN tab

Features:

- Improved tooltip look

## Version 2.1.7 (2020-04-16)

Features:

- Improved documentation and made charts responsive

## Version 2.1.6 (2020-04-09)

Fix:

- Removed old vertical temperature gradient calculator

## Version 2.1.5 (2020-04-06)

Fix:

- Local discomfort tab the code was broken

## Version 2.1.4 (2020-03-30)

Fix:

- Spinner behavior compare tab
- Units in note under range chart
- dilation coefficient in SET equation

Features:

- Implemented test for all the major functions
- Added `bumpversion` to track version of the tool

## Version 2.1.3 (2020-03-17)

Features:

- Changed compliance from ASHRAE 55 2017 to ASHRAE 55 2020

## Version 2.1.2 (2020-03-11)

Fix:

- Equation to calculate heat transfer coefficient SET equation.
- Default value of clo and met dropdown now matches the default input value.

Features:

- Wrote testing for ERF, SET and PMV equations

## Version 2.1.1 (2021-02-04)

Fix:

- Error calculation SHARP and solar altitude supine person.  

## Version 2.1.0 (2020-10-19)

Features:

- Changed limit for elevated air speed calculation from 0.2 m/s to 0.1 m/s.  

## Version 2.0.6 (2020-09-08)

Features:

- Fixed minor issue in the erf equation. Updated fp tables and E_diff equation.  

## Version 2.0.5 (2020-09-04)

Features:

- Fixed issues with calculation of outputs in the Upload tool and removed link to Codelab
- Added SET chart, which displays SET outputs as a function of the inout variables selected
- Added a note below SET chart
- Added a note below PMV heat loss chart
- Minor edits to the layout of the comfort tool  

## Version 2.0.4 (2020-07-17)

Features:

- Displaying the relative air speed and dynamic clothing among the outputs.
  
Fix:

- Fixed calculation comfort zone in Ranges

## Version 2.0.3 (2020-06-24)

Features:

- The Upload tool can be used to calculate LEED compliance.
- The Share button now also shares the chart type.
- Added notes about relative air speed and dynamic clothing calculation below charts.
- Added mean radiant temperature as input in the SolarCal tool.
  
Fix:

- Share fixed the issue with air speed and clothing.
- Changes y-label in chart air speed vs operative temperature.

## Version 2.0.2 (2020-06-09)

Features:

- The [dynamic clothing insulation](http://centerforthebuiltenvironment.github.io/comfort_tool/docs/pmv#dynamic-clothing-insulation) is automatically calculated by the tool.
  
Fix:

- Upload was throwing error for files with more than 1000 inputs.
- Improved calculation speed for Upload tool.

## Version 2.0.1 (2020-06-05)

Features:

- The [relative air speed](http://centerforthebuiltenvironment.github.io/comfort_tool/docs/pmv#relative-air-speed) is automatically calculated by the tool.
- Added Save, Restore and Share button.
- Added tooltips to provide additional information to the user.
  
Fix:

- Upload not does not throw an error if the cooling effect cannot be calculated.
- Upgraded to new version of pythermalcomfort.
- Fixed issue solar gain calculation with IP units.

## Version 2.0.0 (2020-05-06)

This version complies with the ASHRAE Standard 55-2017. You can access this version of the tool using [this link](https://comfort-tool-63o6tweb2a-uc.a.run.app). Please note that this version is no longer supported and may contain errors that have been fixed in newer releases.

Features:

- Navigation bar responsive.
- Created new documentation website.
  
Fix:

- Issue with HeatLoss chart.
  
  

static/js/EN/EN.js
keys = ["ta", "tr", "vel", "rh", "met", "clo", "trm", "vel_a"];

const vRelativeValue = $("#relative-air-speed-value");
const vRelativeDiv = $("#relative-air-speed-div");

$(document).ready(function () {
  // highlight navigation bar button
  $("a.active").removeClass("active");
  $("#nav_a_en").addClass("active");

  var cloSelect = document.getElementById("cloSelect");
  cloSelect.onchange = function () {
    document.getElementById("clo").value = cloSelect.value;
    update();
  };

  populate_clo_dropdown();

  var cloMultiSelect = document.getElementById("cloMultiSelect");
  cloInsulationGarments.forEach(function (element) {
    cloMultiSelect.options.add(new Option(element.article, element.clo));
  });

  const actSelect = document.getElementById("actSelect");
  actSelect.onchange = function () {
    document.getElementById("met").value = actSelect.value;
    update();
  };

  populate_met_dropdown(metRatesTypicalTasks);

  $(function () {
    $(".multiselect").multiselect({
      sortable: false,
      searchable: false,
      dividerLocation: 0.5,
    });
  });

  var velaSelect = document.getElementById("vel_a");

  velaSelect.onchange = function () {
    update();
    var coolingEffect;
    if (d.vel_a === 0.3) {
      coolingEffect = 0;
    } else if (d.vel_a === 0.6) {
      coolingEffect = 1.2;
    } else if (d.vel_a === 0.9) {
      coolingEffect = 1.8;
    } else if (d.vel_a === 1.2) {
      coolingEffect = 2.2;
    }
    ac.redrawBounds(coolingEffect);
  };

  $(
    "#adaptive-inputs, #adaptive-note, #psychtop-note, #temphum-note, #chart-div-adaptive, #temphumchart-div"
  ).hide();
  window.isCelsius = true;
  window.humUnit = "rh";
  resetDefaultValues();
  update();
  bc.drawChart();
  bc.findComfortBoundary(d, 0.5);
  enbc.drawComfortRegions(d);
  bc.drawPoint();

  pc.drawChart();
  var json = [
    {
      db: d.ta,
      hr: pc.getHumRatio(d.ta, d.rh),
    },
  ];
  enpc.drawComfortRegions(d);
  pc.drawPoint(json);

  ac.drawChart();
  ac.drawPoint([d]);
});

$(function () {
  $("#globedialog").dialog({
    autoOpen: false,
    height: 350,
    width: 400,
    modal: true,
    resizable: false,
    buttons: {
      "Set mean radiant temperature": function () {
        var tr = parseFloat($("#mrt-result").val());
        if (!isCelsius) tr = util.CtoF(tr);
        $("#tr").val(tr);
        $(this).dialog("close");
        update();
      },
    },
  });

  $("#localdialog").dialog({
    autoOpen: false,
    height: 600,
    width: 470,
    modal: true,
    resizable: false,
  });

  $("#link").click(function () {
    if ($("#tr-input").is(":hidden")) {
      $("#ta-lab").html(
        '<a class="mainlink" href="http://en.wikipedia.org/wiki/Dry-bulb_temperature" target="_new">Air temperature</a>'
      );
      $("#globeTemp").removeAttr("disabled");
      $("#tr-input, #tr-lab").show();
    } else {
      $("#ta-lab").html(
        '<a class="mainlink" href="http://en.wikipedia.org/wiki/Operative_temperature" target="_new">Operative temperature</a>'
      );
      $("#globeTemp").attr("disabled", "disabled");
      $("#tr-input, #tr-lab").hide();
    }
  });

  $("#radio").buttonset();

  $("#customClo").click(function () {
    $("#customCloToggle").toggle("fast");
  });

  $("button").button();

  $(".buttons").buttonset();

  // create spinners and impose limits based on value defined in envValLimits
  $("#ta").spinner({
    step: envVarLimits.ta.si.step,
    min: envVarLimits.ta.si.min,
    max: envVarLimits.ta.si.max,
    numberFormat: "n",
  });

  $("#tr").spinner({
    step: envVarLimits.tr.si.step,
    min: envVarLimits.tr.si.min,
    max: envVarLimits.tr.si.max,
    numberFormat: "n",
  });

  $("#trm").spinner({
    step: envVarLimits.trm.si.step,
    min: envVarLimits.trm.si.min,
    max: envVarLimits.trm.si.max,
    numberFormat: "n",
  });

  $("#vel").spinner({
    step: envVarLimits.vel.si.step,
    min: envVarLimits.vel.si.min,
    max: envVarLimits.vel.si.max,
    numberFormat: "n",
  });

  $("#clo").spinner({
    step: envVarLimits.clo.step,
    min: envVarLimits.clo.min,
    max: envVarLimits.clo.max,
    numberFormat: "n",
  });

  $("#met").spinner({
    step: envVarLimits.met.step,
    min: envVarLimits.met.min,
    max: envVarLimits.met.max,
    numberFormat: "n",
  });

  $("#rh").spinner({
    step: envVarLimits.rh.step,
    min: envVarLimits.rh.min,
    max: envVarLimits.rh.max,
    numberFormat: "n",
  });
});

$("#humidity-spec").change(function () {
  change_humidity_selection();
});

$("#link").click(function () {
  $("#tr").val($("#ta").val());
});

$(".inputbox").keydown(function (event) {
  if (event.keyCode === 13) {
    var inputs = $(".inputbox:visible:enabled");
    var nextBox = inputs.index(this) + 1;
    if (nextBox === inputs.length) nextBox = 0;
    inputs[nextBox].focus();
  }
});

$(".in").click(function () {
  update();
});

$(".inputbox").focusout(function () {
  update();
});

// $('#vel_a').focusout(function () {
//     update();
// });

$("#vel-a-box").click(function () {
  update();
});

$("#unitsToggle").click(function () {
  toggleUnits();
  update();
});

$("#setDefaults").click(function () {
  resetDefaultValues();
  update();
});

$("#specPressure").click(function () {
  var customPressure = prompt(
    "Enter atmospheric pressure in ".concat(
      isCelsius
        ? "Pascals (Pa) default value 101325 Pa"
        : "inches of mercury (inHg)"
    )
  );
  if (customPressure !== "" && customPressure !== null) {
    customPressure = parseFloat(customPressure);
    if (!isCelsius) {
      customPressure *= 3386.39;
    }
    if (
      !isNaN(customPressure) &&
      customPressure >= 30000 &&
      customPressure <= 110000
    ) {
      psy.PROP.Patm = customPressure;
      enpc.redraw_rh_lines();
      update();
    } else {
      window.alert("The entered atmospheric pressure is invalid.");
    }
  }
});

$("#globeTemp").click(function () {
  var container = $("#globedialog");
  $.ajax({
    url: util.STATIC_URL + "/html/globetemp.html",
    success: function (data) {
      $("#globedialog").html(data);
      if (!isCelsius) {
        $("#ta-g").val("77");
        $("#vel-g").val("20");
        $("#tglobe").val("77");
        $("#diameter").val("6");
        $("#g-ta-unit").html(" &deg;F");
        $("#g-vel-unit").html(" fpm");
        $("#g-tglobe-unit").html(" &deg;F");
        $("#g-globediam-unit").html(" in");
        $("#g-mrt-unit").html(" &deg;F");
      }
    },
    async: false,
  });
  container.dialog("open");
  updateGlobe();
  $(".input-dialog").focusout(function () {
    updateGlobe();
  });
});

$("#localDisc").click(function () {
  var container = $("#localdialog");
  $.ajax({
    url: util.STATIC_URL + "/html/localdisc.html",
    success: function (data) {
      $("#localdialog").html(data);
      if (!isCelsius) {
        $(".tempunit").html(" &deg;F");
        $(".velunit").html(" fpm");
        $("#T_head").val("77");
        $("#T_ankle").val("77");
        $("#T_floor").val("77");
        $("#T_op").val("77");
        $("#local_vel").val("20");
      }
    },
    async: false,
  });
  container.dialog("open");
  $(".input-dialog-local").focusout(function () {
    updateLocalDisc();
  });
});

$("#setClo").click(function () {
  setClo();
  update();
});

$("#addToEnsembles").click(function () {
  addToEnsembles();
});

$("#model-type").change(function () {
  $("#pmv-out-label").html("PMV");
  $("#localDisc").removeAttr("disabled");
  const model = $("#model-type").val();
  if (model === "pmv") {
    $(
      "#pmv-inputs, #pmv-outputs, #cloInput, #actInput, #humidity-spec-cont, #chart-div, #chartSelect-cont, #pmv-notes"
    ).show();
    $(
      "#adaptive-note, #adaptive-inputs, #adaptive-outputs, #chart-div-adaptive, #chart-title-adaptive, #temphumchart-div, #temphumchart-title"
    ).hide();
  } else if (model === "adaptiveComfort") {
    $("#pmv-inputs, #pmv-outputs, #cloInput").hide();
    $(
      "#actInput, #humidity-spec-cont, #chart-div, #temphumchart-div, #pmv-notes, #chartSelect-cont"
    ).hide();
    $(
      "#adaptive-note, #adaptive-inputs, #adaptive-outputs, #chart-div-adaptive, #chart-title-adaptive"
    ).show();
    $("#localDisc").attr("disabled", "disabled");
  }
  update();
});

/* Code used to select which fields to be displayed when the user selct the method: adaptive or PMV */
$("#chartSelect").change(function () {
  const chart = $("#chartSelect").val();

  // PMV method selection
  if (chart === "psychta" || chart === "psychtop") {
    $("#chart-div").show();
    $("#temphumchart-div").hide();

    // Selction to use air temperature as well as mean radiant temperature
    if (chart === "psychta") {
      $("#psychta-note").show();
      $("#psychtop-note, #temphum-note").hide();

      $("#db-axis-C-label").text("Dry-bulb Temperature [°C]");
      $("#db-axis-F-label").text("Dry-bulb Temperature [°F]");

      if ($("#link").is(":checked")) {
        $("#labelforlink").show();
      } else {
        $("#ta-lab").html(
          '<a class="mainlink" href="http://en.wikipedia.org/wiki/Dry-bulb_temperature" target="_new">Air temperature</a>'
        );
        $("#globeTemp").removeAttr("disabled");
        $("#tr-input, #tr-lab, #labelforlink").show();
      }

      // Selction to use air operative temperature
    } else if (chart === "psychtop") {
      $("#psychtop-note").show();
      $("#psychta-note, #temphum-note").hide();

      $("#db-axis-C-label").text("Operative Temperature [°C]");
      $("#db-axis-F-label").text("Operative Temperature [°F]");

      $("#ta-lab").html(
        '<a class="mainlink" href="http://en.wikipedia.org/wiki/Operative_temperature" target="_new">Operative temperature</a>'
      );
      $("#globeTemp").attr("disabled", "disabled");
      $("#tr-input, #tr-lab, #labelforlink").hide();
    }

    // selection Adaptive method
  } else if (chart === "temphum") {
    $("#temphumchart-div, #temphum-note").show();
    $("#chart-div, #psychta-note, #psychtop-note").hide();
    if ($("#link").is(":checked")) {
      $("#labelforlink").show();
    } else {
      $("#ta-lab").html(
        '<a class="mainlink" href="http://en.wikipedia.org/wiki/Dry-bulb_temperature" target="_new">Air temperature</a>'
      );
      $("#globeTemp").removeAttr("disabled");
      $("#tr-input, #tr-lab, #labelforlink").show();
    }
  }
  update();
});

function update() {
  let r;

  if ($("#link").is(":checked") || $("#chartSelect").val() === "psychtop") {
    $("#tr").val($("#ta").val());
  }

  // get user input and validate that complies with standard applicability limits
  validateUserEntry("");

  d.wme = 0;
  if (!isCelsius) {
    d.ta = util.FtoC(d.ta);
    d.tr = util.FtoC(d.tr);
    d.trm = util.FtoC(d.trm);
    d.vel /= 196.9;
    d.vel_a /= 196.9;
    if (window.humUnit === "wetbulb" || window.humUnit === "dewpoint")
      d.rh = util.FtoC(d.rh);
    else if (window.humUnit === "vappress") d.rh *= 2953;
  } else {
    if (window.humUnit === "vappress") d.rh *= 1000;
  }

  // calculate and display relative air speed
  if (d.met > 1) {
    vRelativeDiv.show();
    if (isCelsius) {
      vRelativeValue.html(comf.relativeAirSpeed(d.vel, d.met).toFixed(2));
    } else {
      vRelativeValue.html(
        (comf.relativeAirSpeed(d.vel, d.met) * 196).toFixed(2)
      );
    }
  } else {
    vRelativeDiv.hide();
  }

  const model = document.getElementById("model-type").value;
  if (model === "pmv") {
    r = comf.pmv(
      d.ta,
      d.tr,
      comf.relativeAirSpeed(d.vel, d.met),
      d.rh,
      d.met,
      comf.dynamicClothing(d.clo, d.met),
      0
    );
    if (isNaN(r.pmv)) {
      window.alert(
        "The combination of input parameters you selected lead to an incorrect calculation of the PMV index\n" +
          "Please check that the value you entered are correct.\n" +
          "The input parameters has been set back to their default values."
      );
      resetDefaultValues();
    }
    renderPmvResults(r);
    calcPmvCompliance(d, r);
    if ($("#chart-div").is(":visible")) {
      enpc.redrawComfortRegions(d);
      var pointdata = [
        {
          db: d.ta,
          hr: pc.getHumRatio(d.ta, d.rh),
        },
      ];
      pc.redrawPoint(pointdata);
    } else if ($("#temphumchart-div").is(":visible")) {
      enbc.redrawComfortRegions(d);
      bc.redrawPoint();
    }
  } else if (model === "adaptiveComfort") {
    r = comf.adaptiveComfortEN(d.ta, d.tr, d.trm, d.vel_a);
    renderAdaptiveResults(r);
    calcAdaptiveCompliance(d, r);
    ac.redrawPoint([d]);
  }
}

function getCategory(pmv) {
  if (Math.abs(pmv) <= 0.2) return "I";
  else if (Math.abs(pmv) <= 0.5) return "II";
  else if (Math.abs(pmv) <= 0.7) return "III";
  else if (Math.abs(pmv) > 0.7) return "IV";
  else return "-";
}

function renderPmvResults(r) {
  $("#pmv-res").html(r.pmv.toFixed(2));
  $("#ppd-res").html(r.ppd.toFixed(0));
  var category = getCategory(r.pmv);
  $("#category").html(category);
}

function renderAdaptiveResults(r) {
  const to = (parseFloat($("#ta").val()) + parseFloat($("#tr").val())) / 2;
  if (!isCelsius) {
    r.tComfIUpper = util.CtoF(r.tComfIUpper);
    r.tComfIIUpper = util.CtoF(r.tComfIIUpper);
    r.tComfIIIUpper = util.CtoF(r.tComfIIIUpper);
    r.tComfILower = util.CtoF(r.tComfILower);
    r.tComfIILower = util.CtoF(r.tComfIILower);
    r.tComfIIILower = util.CtoF(r.tComfIIILower);
  }
  $("#limitsI").html(
    "Operative temperature: " +
      r.tComfILower.toFixed(1) +
      " to " +
      r.tComfIUpper.toFixed(1)
  );
  $("#limitsII").html(
    "Operative temperature: " +
      r.tComfIILower.toFixed(1) +
      " to " +
      r.tComfIIUpper.toFixed(1)
  );
  $("#limitsIII").html(
    "Operative temperature: " +
      r.tComfIIILower.toFixed(1) +
      " to " +
      r.tComfIIIUpper.toFixed(1)
  );
  if (r.acceptabilityI) {
    $("#sensationIII, #sensationII, #sensationI").html("Comfortable");
  } else if (r.acceptabilityII) {
    $("#sensationIII, #sensationII").html("Comfortable");
    if (to < r.tComfIUpper) {
      $("#sensationI").html('<span style="color:blue;">Too cool</span>');
    } else {
      $("#sensationI").html('<span style="color:red;">Too warm</span>');
    }
  } else if (r.acceptabilityIII) {
    $("#sensationIII").html("Comfortable");
    if (to < r.tComfIIUpper) {
      $("#sensationI, #sensationII").html(
        '<span style="color:blue;">Too cool</span>'
      );
    } else {
      $("#sensationI, #sensationII").html(
        '<span style="color:red;">Too warm</span>'
      );
    }
  } else if (to < r.tComfIIIUpper) {
    $("#sensationIII, #sensationII, #sensationI").html(
      '<span style="color:blue;">Too cool</span>'
    );
  } else {
    $("#sensationIII, #sensationII, #sensationI").html(
      '<span style="color:red;">Too warm</span>'
    );
  }
}

function calcPmvCompliance(d, r) {
  const pmv_complyIII = Math.abs(r.pmv) <= 0.7;
  const met_comply = d.met <= 4 && d.met >= 0.8;
  const clo_comply = d.clo <= 2;

  var special_msg = "";
  let comply = true;

  if (!met_comply) {
    comply = false;
    special_msg +=
      "Metabolic rates below 0.8 or above 4.0 are not covered by this standard<br>";
  }
  if (!clo_comply) {
    comply = false;
    special_msg += "Clo values above 2.0 are not covered by this standard<br>";
  }
  if (!pmv_complyIII) {
    comply = false;
  }

  renderCompliance(comply, special_msg);
}

function calcAdaptiveCompliance(d, r) {
  var comply = true;
  var special_msg = "";

  if (d.trm > 30 || d.trm < 10) {
    comply = false;
    special_msg +=
      "&#8627; Running mean outdoor temperatures above " +
      (isCelsius ? "30&deg;C " : "92.3&deg;F ") +
      "or below " +
      (isCelsius ? "10&deg;C " : "50&deg;F ") +
      "are not covered by Standard-55<br>";
  }
  if (!r.acceptabilityIII) comply = false;

  renderCompliance(comply, special_msg);
}

function renderCompliance(comply, special_msg) {
  const comply_msg = "&#10004; &nbsp;Complies with EN-16798";
  const no_comply_msg = "&#10008 &nbsp; Does not comply with EN-16798";

  $("#vel-range").html("");
  if (comply) {
    $("#comply-msg").html(comply_msg);
    $("#output-b")
      .removeClass("alert alert-danger")
      .addClass("alert alert-success")
      .css({ color: "green" });
    $("#special-msg").html(special_msg);
  } else {
    $("#comply-msg").html(no_comply_msg);
    $("#output-b")
      .removeClass("alert alert-success")
      .addClass("alert alert-danger")
      .css({ color: "red" });
    $("#special-msg").html(special_msg);
  }
}

// Set clo value created by the custom ensemble dialog
function setClo() {
  let clo = 0;
  var opt = document.getElementById("cloMultiSelect").options;
  for (var i = 0; i < opt.length; i++) {
    if (opt[i].selected) clo += parseFloat(opt[i].value);
  }
  document.getElementById("clo").value = clo.toFixed(2);
}

// Add selected clothing items to create a custom ensemble
function addToEnsembles() {
  var items = [];
  var ensembleClo = 0;
  var opt = document.getElementById("cloMultiSelect").options;
  for (var i = 0; i < opt.length; i++) {
    if (opt[i].selected) {
      items.push(opt[i].text);
      ensembleClo += parseFloat(opt[i].value);
    }
  }
  cloSelect.options.add(new Option(items.join(", "), ensembleClo.toFixed(2)));
}

function updateGlobe() {
  let ta = parseFloat($("#ta-g").val());
  let vel = parseFloat($("#vel-g").val());
  let tglobe = parseFloat($("#tglobe").val());
  let diameter = parseFloat($("#diameter").val());
  let emissivity = parseFloat($("#emissivity").val());
  if (!isCelsius) {
    ta = util.FtoC(ta);
    vel /= 196.9;
    tglobe = util.FtoC(tglobe);
    diameter *= 0.0254;
  }
  let tr = psy.globetemp(ta, vel, tglobe, diameter, emissivity);
  if (!isCelsius) tr = util.CtoF(tr);
  $("#mrt-result").val(tr.toFixed(1));
}
